cache/src/lib.rs

```diff
@@ -150,46 +155,156 @@ pub struct CacheItem {
-}
-impl TryFrom<atspi::cache::CacheItem> for CacheItem {
-	type Error = AccessiblePrimitiveConversionError;
+	// The children (ids) of the accessible.
+	pub children: Vec<AccessiblePrimitive>,
 
-	fn try_from(atspi_cache_item: atspi::cache::CacheItem) -> Result<Self, Self::Error> {
+	#[serde(skip)]
+	pub cache: Weak<Cache>,
+}
+impl CacheItem {
+	pub async fn from_atspi_event<T: Signified>(event: &T, cache: Arc<Cache>, connection: &zbus::Connection) -> OdiliaResult<Self> {
+		let a11y_prim = AccessiblePrimitive::from_event(event)?;
+		accessible_to_cache_item(
+			&a11y_prim.into_accessible(connection).await?,
+			cache
+		).await
+	}
+	pub async fn from_atspi_cache_item(atspi_cache_item: atspi::cache::CacheItem, cache: Arc<Cache>, connection: &zbus::Connection) -> OdiliaResult<Self> {
+		let children_primitives: Vec<AccessiblePrimitive> = AccessiblePrimitive::try_from(atspi_cache_item.object.clone())?
+			.into_accessible(connection)
+			.await?
+			.get_children()
+			.await?
+			.into_iter()
+			.map(|child_object_pair| child_object_pair.try_into())
+			.collect::<Result<Vec<AccessiblePrimitive>, AccessiblePrimitiveConversionError>>()?;
 		Ok(Self {
 			object: atspi_cache_item.object.try_into()?,
 			app: atspi_cache_item.app.try_into()?,
 			parent: atspi_cache_item.parent.try_into()?,
 			index: atspi_cache_item.index,
-			children: atspi_cache_item.children,
-			ifaces: atspi_cache_item.ifaces,
+			children_num: atspi_cache_item.children,
+			interfaces: atspi_cache_item.ifaces,
 			role: atspi_cache_item.role,
 			states: atspi_cache_item.states,
 			text: atspi_cache_item.name,
+			cache: Arc::downgrade(&cache),
+			children: children_primitives,
 		})
 	}
 }
 
-/// The root of the accessible cache.
-#[derive(Clone)]
-pub struct Cache {
-	pub by_id: Arc<RwLock<HashMap<AccessibleId, CacheItem>>>,
+macro_rules! strong_cache {
+	($cache_ref:expr) => {
+		Weak::upgrade($cache_ref)
+			.ok_or(OdiliaError::Cache(CacheError::NotAvailable))?
+	}
 }
-// clippy wants this
-impl Default for Cache {
-	fn default() -> Self {
-		Self::new()
+macro_rules! as_accessible {
+	($self:expr) => {
+		{
+			let cache = strong_cache!(&$self.cache);
+			$self.object.clone().into_accessible(&cache.connection).await?
+		}
 	}
 }
 
+#[async_trait]
+impl Accessible for CacheItem {
+	type Error = OdiliaError;
+
+	async fn get_application(&self) -> Result<Self, Self::Error> {
+		let derefed_cache: Arc<Cache> = strong_cache!(&self.cache);
+		derefed_cache.get(&self.app).await.ok_or(CacheError::NoItem.into())
+	}
+	async fn parent(&self) -> Result<Self, Self::Error> {
+		let derefed_cache: Arc<Cache> = strong_cache!(&self.cache);
+		derefed_cache.get(&self.parent).await.ok_or(CacheError::NoItem.into())
+	}
+	async fn get_children(&self) -> Result<Vec<Self>, Self::Error> {
+		let derefed_cache: Arc<Cache> = strong_cache!(&self.cache);
+		derefed_cache.get_all(&self.children).await
+			.into_iter()
+			.map(|child| child.ok_or(CacheError::NoItem.into()))
+			.collect()
```

Comment:
In general, I'm just trying to fail hard so that we can actually find problems. If I just filter out chlidren, then we could end up with very strange and hard to debug issues.

Much better to fail loudly.